composer/utils/checkpoint.py

```diff
@@ -200,46 +201,158 @@ def load_checkpoint(
-    needs_unique_checkpoint_folder = state.fsdp_sharded_state_dict_enabled or dist.get_local_rank() == 0
-    tempdir_ctx = tempfile.TemporaryDirectory() if needs_unique_checkpoint_folder else contextlib.nullcontext(None)
-    with tempdir_ctx as tempdir:
-        try:
-            # Get the path to the proper checkpoint folder corresponding to the current rank's node.
-            # If fsdp_sharded_state_dict_enabled then just use that rank's unique tempdir.
-            node_checkpoint_folder = (tempdir
-                                      if state.fsdp_sharded_state_dict_enabled else _get_local_rank_zero_path(tempdir))
-            assert node_checkpoint_folder is not None
-
-            composer_states_filepath, extracted_checkpoint_folder, extracted_rank_n = download_checkpoint(
-                path=path,
-                node_checkpoint_folder=node_checkpoint_folder,
-                object_store=object_store,
-                progress_bar=progress_bar,
-                fsdp_sharded_state_dict_enabled=state.fsdp_sharded_state_dict_enabled,
-                deepspeed_sharded_checkpoint=is_model_deepspeed(state.model),
-            )
-            rng_state_dicts = _restore_checkpoint(
-                state,
-                logger,
-                composer_states_filepath,
-                extracted_rank_n,
-                extracted_checkpoint_folder,
-                load_weights_only=load_weights_only,
-                strict_model_weights=strict_model_weights,
-                ignore_keys=ignore_keys,
-                exclude_algorithms=exclude_algorithms,
-                algorithm_passes=algorithm_passes,
-            )
-        finally:
-            # Wait for all ranks to finish restoring the checkpoint before releasing the tempdir, since tempdir can
-            # be a shared resource between nodes.
-            dist.barrier()
-
-    log.info('%s loaded from %s', 'Model weights' if load_weights_only else 'Trainer checkpoint', path)
+    if state.fsdp_sharded_state_dict_enabled:
+        rng_state_dicts = load_sharded_checkpoint(
+            source_path=path,
+            state=state,
+            logger=logger,
+            object_store=object_store,
+            load_weights_only=load_weights_only,
+            strict_model_weights=strict_model_weights,
+            progress_bar=progress_bar,
+            ignore_keys=ignore_keys,
+            exclude_algorithms=exclude_algorithms,
+            algorithm_passes=algorithm_passes,
+        )
+    else:
+        # Download the checkpoint to the node-local folder
+        log.debug('Loading checkpoint at %s', path)
+        # Each node gets one unique folder to store checkpoints that is shared amongst all local ranks in that node.
+        # If fsdp sharded state_dicts is enabled then EVERY rank gets a unique checkpoint folder.
+        needs_unique_checkpoint_folder = state.fsdp_sharded_state_dict_enabled or dist.get_local_rank() == 0
```

Comment:
seems like all the sharded bits in this else should be removed from this code path since it is handled in the if above?

Reply:
maybe keep it if someone uses torch <2.0 to do sharding?

Reply:
but the if above is `if state.fsdp_sharded_state_dict_enabled:`. so you can't end up in this code path?

Reply:
is the if statement above supposed to gate on torch 2?

Reply:
oh good call yeah should change to gate by torch version as well

Reply:
ok done